File: nodejs/src/ledger.ts
import Database from "better-sqlite3";
import { randomBytes } from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { EventType, type ExecutionEvent } from "./events.js";
import { sha256Hex, stableJsonDumps } from "./hashing.js";
import { assertTransition } from "./stateMachine.js";
import { SqubeApprovalError } from "./errors.js";
import { ActionStatus, type ActionRecord } from "./types.js";

export interface ApprovalRecord {
  approval_id: string;
  execution_id: string;
  status: string;
  parameters_hash: string;
  requested_at: string;
  expires_at: string | null;
  decided_at: string | null;
  decided_by: string | null;
  decision_reason: string | null;
}

const SCHEMA = `
CREATE TABLE IF NOT EXISTS execution_records (
  execution_id TEXT PRIMARY KEY,
  created_at TEXT NOT NULL,
  agent_id TEXT NOT NULL,
  action TEXT NOT NULL,
  resource TEXT,
  parameters_hash TEXT NOT NULL,
  parameters_summary TEXT,
  decision TEXT NOT NULL,
  policy_id TEXT NOT NULL,
  status TEXT NOT NULL,
  approved_by TEXT,
  approval_reason TEXT,
  result_status TEXT,
  error_message TEXT,
  duration_ms INTEGER,
  completed_at TEXT
);

CREATE TABLE IF NOT EXISTS execution_events (
  event_id TEXT PRIMARY KEY,
  execution_id TEXT NOT NULL,
  event_type TEXT NOT NULL,
  timestamp TEXT NOT NULL,
  actor TEXT NOT NULL,
  payload TEXT NOT NULL,
  previous_event_hash TEXT,
  event_hash TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS approval_requests (
  approval_id TEXT PRIMARY KEY,
  execution_id TEXT NOT NULL,
  status TEXT NOT NULL,
  parameters_hash TEXT NOT NULL,
  requested_at TEXT NOT NULL,
  expires_at TEXT,
  decided_at TEXT,
  decided_by TEXT,
  decision_reason TEXT
);
`;

function newEventId(): string {
  return `sq_evt_${randomBytes(8).toString("hex")}`;
}

export class Ledger {
  private db: Database.Database;

  constructor(ledgerPath = "sqube_ledger.sqlite3") {
    const dir = path.dirname(ledgerPath);
    if (dir && dir !== ".") fs.mkdirSync(dir, { recursive: true });
    this.db = new Database(ledgerPath);
    this.db.exec(SCHEMA);
    this.db.pragma("journal_mode = WAL");
    this.migrate();
  }

  private migrate(): void {
    const cols = this.db.prepare("PRAGMA table_info(execution_records)").all() as Array<{
      name: string;
    }>;
    const names = new Set(cols.map((c) => c.name));
    if (!names.has("correlation_id")) {
      this.db.exec("ALTER TABLE execution_records ADD COLUMN correlation_id TEXT");
    }
    if (!names.has("root_execution_id")) {
      this.db.exec("ALTER TABLE execution_records ADD COLUMN root_execution_id TEXT");
    }
  }

  getExecution(executionId: string): ActionRecord | undefined {
    return this.db
      .prepare("SELECT * FROM execution_records WHERE execution_id = ?")
      .get(executionId) as ActionRecord | undefined;
  }

  transitionExecution(
    executionId: string,
    fromStatus: string,
    toStatus: string,
    fields: Partial<ActionRecord> = {}
  ): void {
    assertTransition(fromStatus, toStatus, executionId);
    const keys = Object.keys(fields);
    const sets = ["status = @to_status", ...keys.map((k) => `${k} = @${k}`)];
    const stmt = this.db.prepare(
      `UPDATE execution_records SET ${sets.join(", ")} WHERE execution_id = @execution_id AND status = @from_status`
    );
    const result = stmt.run({
      ...fields,
      to_status: toStatus,
      execution_id: executionId,
      from_status: fromStatus,
    });
    if (result.changes !== 1) {
      const row = this.getExecution(executionId);
      throw new Error(
        `Invalid execution transition ${row?.status ?? "MISSING"} -> ${toStatus}`
      );
    }
  }

  insert(record: ActionRecord): void {
    const stmt = this.db.prepare(`
      INSERT INTO execution_records (
        execution_id, created_at, agent_id, action, resource,
        parameters_hash, parameters_summary, decision, policy_id, status,
        approved_by, approval_reason, result_status, error_message,
        duration_ms, completed_at
      ) VALUES (
        @execution_id, @created_at, @agent_id, @action, @resource,
        @parameters_hash, @parameters_summary, @decision, @policy_id, @status,
        @approved_by, @approval_reason, @result_status, @error_message,
        @duration_ms, @completed_at
      )
    `);
    stmt.run({
      ...record,
      approved_by: record.approved_by ?? null,
      approval_reason: record.approval_reason ?? null,
      result_status: record.result_status ?? null,
      error_message: record.error_message ?? null,
      duration_ms: record.duration_ms ?? null,
      completed_at: record.completed_at ?? null,
    });
  }

  updateStatus(executionId: string, fields: Partial<ActionRecord>): void {
    const keys = Object.keys(fields);
    if (keys.length === 0) return;
    const sets = keys.map((k) => `${k} = @${k}`).join(", ");
    const stmt = this.db.prepare(
      `UPDATE execution_records SET ${sets} WHERE execution_id = @execution_id`
    );
    stmt.run({ ...fields, execution_id: executionId });
  }

  appendEvent(
    executionId: string,
    eventType: EventType,
    actor: string,
    payload: Record<string, unknown>,
    timestamp: string
  ): ExecutionEvent {
    const prev = this.db
      .prepare(
        "SELECT event_hash FROM execution_events WHERE execution_id = ? " +
          "ORDER BY timestamp DESC LIMIT 1"
      )
      .get(executionId) as { event_hash: string } | undefined;
    const previousHash = prev?.event_hash ?? null;
    const eventId = newEventId();
    const body = stableJsonDumps({
      event_id: eventId,
      execution_id: executionId,
      event_type: eventType,
      timestamp,
      actor,
      payload,
      previous_event_hash: previousHash,
    });
    const eventHash = sha256Hex(body);
    this.db
      .prepare(
        `INSERT INTO execution_events (
          event_id, execution_id, event_type, timestamp, actor, payload,
          previous_event_hash, event_hash
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)`
      )
      .run(
        eventId,
        executionId,
        eventType,
        timestamp,
        actor,
        stableJsonDumps(payload),
        previousHash,
        eventHash
      );
    return {
      event_id: eventId,
      execution_id: executionId,
      event_type: eventType,
      timestamp,
      actor,
      payload,
      previous_event_hash: previousHash,
      event_hash: eventHash,
    };
  }

  getEvents(executionId: string): ExecutionEvent[] {
    const rows = this.db
      .prepare(
        "SELECT * FROM execution_events WHERE execution_id = ? ORDER BY timestamp ASC"
      )
      .all(executionId) as Array<{
      event_id: string;
      execution_id: string;
      event_type: string;
      timestamp: string;
      actor: string;
      payload: string;
      previous_event_hash: string | null;
      event_hash: string;
    }>;
    return rows.map((row) => ({
      event_id: row.event_id,
      execution_id: row.execution_id,
      event_type: row.event_type as EventType,
      timestamp: row.timestamp,
      actor: row.actor,
      payload: JSON.parse(row.payload) as Record<string, unknown>,
      previous_event_hash: row.previous_event_hash,
      event_hash: row.event_hash,
    }));
  }

  verifyChain(executionId?: string): boolean {
    if (executionId) {
      const events = this.getEvents(executionId);
      for (const event of events) {
        const body = stableJsonDumps({
          event_id: event.event_id,
          execution_id: event.execution_id,
          event_type: event.event_type,
          timestamp: event.timestamp,
          actor: event.actor,
          payload: event.payload,
          previous_event_hash: event.previous_event_hash,
        });
        if (sha256Hex(body) !== event.event_hash) return false;
      }
      return true;
    }
    const ids = this.db
      .prepare("SELECT DISTINCT execution_id FROM execution_events")
      .all() as Array<{ execution_id: string }>;
    if (ids.length === 0) return true;
    return ids.every((row) => this.verifyChain(row.execution_id));
  }

  listExecutions(limit = 50, status?: string): ActionRecord[] {
    if (status) {
      return this.db
        .prepare(
          "SELECT * FROM execution_records WHERE status = ? ORDER BY created_at DESC LIMIT ?"
        )
        .all(status, limit) as ActionRecord[];
    }
    return this.db
      .prepare("SELECT * FROM execution_records ORDER BY created_at DESC LIMIT ?")
      .all(limit) as ActionRecord[];
  }

  getLatest(): ActionRecord | undefined {
    const row = this.db
      .prepare("SELECT * FROM execution_records ORDER BY created_at DESC LIMIT 1")
      .get() as ActionRecord | undefined;
    return row;
  }

  createApprovalRequest(
    approvalId: string,
    executionId: string,
    parametersHash: string,
    requestedAt: string,
    expiresAt: string | null
  ): void {
    this.db
      .prepare(
        `INSERT INTO approval_requests (
          approval_id, execution_id, status, parameters_hash, requested_at, expires_at
        ) VALUES (?, ?, 'PENDING', ?, ?, ?)`
      )
      .run(approvalId, executionId, parametersHash, requestedAt, expiresAt);
  }

  getApproval(approvalId: string): ApprovalRecord | undefined {
    return this.db
      .prepare("SELECT * FROM approval_requests WHERE approval_id = ?")
      .get(approvalId) as ApprovalRecord | undefined;
  }

  grantApproval(approvalId: string, decidedBy: string, nowIso: string): string {
    const grant = this.db.transaction(() => {
      const row = this.getApproval(approvalId);
      if (!row) throw new SqubeApprovalError(approvalId, "not found");
      if (row.status !== "PENDING") {
        throw new SqubeApprovalError(approvalId, `status is ${row.status}`);
      }
      if (row.expires_at && row.expires_at < nowIso) {
        throw new SqubeApprovalError(approvalId, "expired");
      }
      const cur = this.db
        .prepare(
          "UPDATE approval_requests SET status = 'GRANTED', decided_at = ?, decided_by = ? WHERE approval_id = ? AND status = 'PENDING'"
        )
        .run(nowIso, decidedBy, approvalId);
      if (cur.changes !== 1) {
        throw new SqubeApprovalError(approvalId, "concurrent decision lost race");
      }
      this.transitionExecution(
        row.execution_id,
        ActionStatus.WAITING_APPROVAL,
        ActionStatus.APPROVED,
        { approved_by: decidedBy }
      );
      return row.execution_id;
    });
    return grant();
  }

  consumeApproval(approvalId: string, executionId: string, nowIso: string): void {
    const consume = this.db.transaction(() => {
      const row = this.getApproval(approvalId);
      if (!row || row.execution_id !== executionId) {
        throw new SqubeApprovalError(approvalId, "not found for execution");
      }
      if (row.status === "CONSUMED") {
        throw new SqubeApprovalError(approvalId, "already consumed");
      }
      if (row.status !== "GRANTED") {
        throw new SqubeApprovalError(approvalId, `status is ${row.status}`);
      }
      const cur = this.db
        .prepare(
          "UPDATE approval_requests SET status = 'CONSUMED', decided_at = ? WHERE approval_id = ? AND status = 'GRANTED'"
        )
        .run(nowIso, approvalId);
      if (cur.changes !== 1) {
        throw new SqubeApprovalError(approvalId, "concurrent consume lost race");
      }
      this.transitionExecution(
        executionId,
        ActionStatus.APPROVED,
        ActionStatus.EXECUTING
      );
    });
    consume();
  }

  close(): void {
    this.db.close();
  }
}

File: rust/src/ledger.rs
use crate::error::{SqubeApprovalError, SqubeGuardError};
use crate::models::{ActionRecord, ActionStatus};
use crate::state_machine::assert_transition;
use rusqlite::{params, Connection, OptionalExtension};
use std::path::Path;

const SCHEMA: &str = "
CREATE TABLE IF NOT EXISTS execution_records (
  execution_id TEXT PRIMARY KEY,
  created_at TEXT NOT NULL,
  agent_id TEXT NOT NULL,
  action TEXT NOT NULL,
  resource TEXT,
  parameters_hash TEXT NOT NULL,
  parameters_summary TEXT,
  decision TEXT NOT NULL,
  policy_id TEXT NOT NULL,
  status TEXT NOT NULL,
  approved_by TEXT,
  approval_reason TEXT,
  result_status TEXT,
  error_message TEXT,
  duration_ms INTEGER,
  completed_at TEXT
);

CREATE TABLE IF NOT EXISTS approval_requests (
  approval_id TEXT PRIMARY KEY,
  execution_id TEXT NOT NULL,
  status TEXT NOT NULL,
  parameters_hash TEXT NOT NULL,
  requested_at TEXT NOT NULL,
  expires_at TEXT,
  decided_at TEXT,
  decided_by TEXT,
  decision_reason TEXT
);
";

pub struct Ledger {
    conn: Connection,
}

impl Ledger {
    pub fn new(path: &str) -> rusqlite::Result<Self> {
        if let Some(parent) = Path::new(path).parent() {
            if !parent.as_os_str().is_empty() {
                std::fs::create_dir_all(parent).map_err(|e| {
                    rusqlite::Error::ToSqlConversionFailure(Box::new(e))
                })?;
            }
        }
        let conn = Connection::open(path)?;
        conn.execute_batch(SCHEMA)?;
        Ok(Self { conn })
    }

    pub fn insert(&self, record: &ActionRecord) -> rusqlite::Result<()> {
        self.conn.execute(
            "INSERT INTO execution_records (
                execution_id, created_at, agent_id, action, resource,
                parameters_hash, parameters_summary, decision, policy_id, status,
                approved_by, approval_reason, result_status, error_message,
                duration_ms, completed_at
            ) VALUES (?1,?2,?3,?4,?5,?6,?7,?8,?9,?10,?11,?12,?13,?14,?15,?16)",
            params![
                record.execution_id,
                record.created_at,
                record.agent_id,
                record.action,
                record.resource,
                record.parameters_hash,
                record.parameters_summary,
                record.decision,
                record.policy_id,
                record.status,
                record.approved_by,
                record.approval_reason,
                record.result_status,
                record.error_message,
                record.duration_ms,
                record.completed_at,
            ],
        )?;
        Ok(())
    }

    pub fn set_status(&self, execution_id: &str, status: &str) -> rusqlite::Result<()> {
        self.conn.execute(
            "UPDATE execution_records SET status = ?1 WHERE execution_id = ?2",
            params![status, execution_id],
        )?;
        Ok(())
    }

    pub fn set_decision(&self, execution_id: &str, decision: &str, status: &str) -> rusqlite::Result<()> {
        self.conn.execute(
            "UPDATE execution_records SET decision = ?1, status = ?2 WHERE execution_id = ?3",
            params![decision, status, execution_id],
        )?;
        Ok(())
    }

    pub fn mark_blocked(&self, execution_id: &str, completed_at: &str) -> rusqlite::Result<()> {
        self.conn.execute(
            "UPDATE execution_records SET status = 'BLOCKED', completed_at = ?1 WHERE execution_id = ?2",
            params![completed_at, execution_id],
        )?;
        Ok(())
    }

    pub fn mark_approval_wait(&self, execution_id: &str) -> rusqlite::Result<()> {
        self.set_status(execution_id, "WAITING_APPROVAL")
    }

    pub fn mark_denied(
        &self,
        execution_id: &str,
        status: &str,
        reason: &str,
        completed_at: &str,
    ) -> rusqlite::Result<()> {
        self.conn.execute(
            "UPDATE execution_records SET status = ?1, approval_reason = ?2, completed_at = ?3 WHERE execution_id = ?4",
            params![status, reason, completed_at, execution_id],
        )?;
        Ok(())
    }

    pub fn mark_approved(&self, execution_id: &str, approved_by: &str) -> rusqlite::Result<()> {
        self.conn.execute(
            "UPDATE execution_records SET status = 'APPROVED', approved_by = ?1 WHERE execution_id = ?2",
            params![approved_by, execution_id],
        )?;
        Ok(())
    }

    pub fn mark_executing(&self, execution_id: &str) -> rusqlite::Result<()> {
        self.set_status(execution_id, "EXECUTING")
    }

    pub fn mark_succeeded(
        &self,
        execution_id: &str,
        duration_ms: i64,
        completed_at: &str,
    ) -> rusqlite::Result<()> {
        self.conn.execute(
            "UPDATE execution_records SET status = 'SUCCEEDED', result_status = 'SUCCESS', duration_ms = ?1, completed_at = ?2 WHERE execution_id = ?3",
            params![duration_ms, completed_at, execution_id],
        )?;
        Ok(())
    }

    pub fn mark_failed(
        &self,
        execution_id: &str,
        error_message: &str,
        duration_ms: Option<i64>,
        completed_at: &str,
    ) -> rusqlite::Result<()> {
        self.conn.execute(
            "UPDATE execution_records SET status = 'FAILED', result_status = 'FAILED', error_message = ?1, duration_ms = ?2, completed_at = ?3 WHERE execution_id = ?4",
            params![error_message, duration_ms, completed_at, execution_id],
        )?;
        Ok(())
    }

    pub fn get_execution(&self, execution_id: &str) -> rusqlite::Result<Option<ActionRecord>> {
        self.conn
            .query_row(
                "SELECT execution_id, created_at, agent_id, action, resource, parameters_hash,
                 parameters_summary, decision, policy_id, status, approved_by, approval_reason,
                 result_status, error_message, duration_ms, completed_at
                 FROM execution_records WHERE execution_id = ?1",
                params![execution_id],
                |row| {
                    Ok(ActionRecord {
                        execution_id: row.get(0)?,
                        created_at: row.get(1)?,
                        agent_id: row.get(2)?,
                        action: row.get(3)?,
                        resource: row.get(4)?,
                        parameters_hash: row.get(5)?,
                        parameters_summary: row.get(6)?,
                        decision: row.get(7)?,
                        policy_id: row.get(8)?,
                        status: row.get(9)?,
                        approved_by: row.get(10)?,
                        approval_reason: row.get(11)?,
                        result_status: row.get(12)?,
                        error_message: row.get(13)?,
                        duration_ms: row.get(14)?,
                        completed_at: row.get(15)?,
                    })
                },
            )
            .optional()
    }

    pub fn transition_execution(
        &self,
        execution_id: &str,
        from: ActionStatus,
        to: ActionStatus,
    ) -> rusqlite::Result<()> {
        assert_transition(from, to).map_err(|e| rusqlite::Error::ToSqlConversionFailure(Box::new(e)))?;
        let n = self.conn.execute(
            "UPDATE execution_records SET status = ?1 WHERE execution_id = ?2 AND status = ?3",
            params![to.as_str(), execution_id, from.as_str()],
        )?;
        if n != 1 {
            return Err(rusqlite::Error::ToSqlConversionFailure(Box::new(
                SqubeGuardError(format!("transition failed {} -> {}", from.as_str(), to.as_str())),
            )));
        }
        Ok(())
    }

    pub fn set_completed(&self, execution_id: &str, completed_at: &str) -> rusqlite::Result<()> {
        self.conn.execute(
            "UPDATE execution_records SET completed_at = ?1 WHERE execution_id = ?2",
            params![completed_at, execution_id],
        )?;
        Ok(())
    }

    pub fn set_denied_meta(
        &self,
        execution_id: &str,
        reason: &str,
        completed_at: &str,
    ) -> rusqlite::Result<()> {
        self.conn.execute(
            "UPDATE execution_records SET approval_reason = ?1, completed_at = ?2 WHERE execution_id = ?3",
            params![reason, completed_at, execution_id],
        )?;
        Ok(())
    }

    pub fn set_approved_by(&self, execution_id: &str, approved_by: &str) -> rusqlite::Result<()> {
        self.conn.execute(
            "UPDATE execution_records SET approved_by = ?1 WHERE execution_id = ?2",
            params![approved_by, execution_id],
        )?;
        Ok(())
    }

    pub fn mark_success_meta(
        &self,
        execution_id: &str,
        duration_ms: i64,
        completed_at: &str,
    ) -> rusqlite::Result<()> {
        self.conn.execute(
            "UPDATE execution_records SET result_status = 'SUCCESS', duration_ms = ?1, completed_at = ?2 WHERE execution_id = ?3",
            params![duration_ms, completed_at, execution_id],
        )?;
        Ok(())
    }

    pub fn create_approval_request(
        &self,
        approval_id: &str,
        execution_id: &str,
        parameters_hash: &str,
        requested_at: &str,
        expires_at: Option<&str>,
    ) -> rusqlite::Result<()> {
        self.conn.execute(
            "INSERT INTO approval_requests (approval_id, execution_id, status, parameters_hash, requested_at, expires_at)
             VALUES (?1, ?2, 'PENDING', ?3, ?4, ?5)",
            params![approval_id, execution_id, parameters_hash, requested_at, expires_at],
        )?;
        Ok(())
    }

    pub fn grant_approval(
        &self,
        approval_id: &str,
        decided_by: &str,
        now_iso: &str,
    ) -> rusqlite::Result<String> {
        let tx = self.conn.unchecked_transaction()?;
        let row = tx
            .query_row(
                "SELECT execution_id, status FROM approval_requests WHERE approval_id = ?1",
                params![approval_id],
                |row| Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?)),
            )
            .optional()?;
        let (execution_id, status) = row.ok_or_else(|| {
            rusqlite::Error::ToSqlConversionFailure(Box::new(SqubeApprovalError {
                approval_id: approval_id.to_string(),
                reason: "not found".into(),
            }))
        })?;
        if status != "PENDING" {
            return Err(rusqlite::Error::ToSqlConversionFailure(Box::new(SqubeApprovalError {
                approval_id: approval_id.to_string(),
                reason: format!("status is {}", status),
            })));
        }
        let n = tx.execute(
            "UPDATE approval_requests SET status = 'GRANTED', decided_at = ?1, decided_by = ?2
             WHERE approval_id = ?3 AND status = 'PENDING'",
            params![now_iso, decided_by, approval_id],
        )?;
        if n != 1 {
            return Err(rusqlite::Error::ToSqlConversionFailure(Box::new(SqubeApprovalError {
                approval_id: approval_id.to_string(),
                reason: "concurrent decision lost race".into(),
            })));
        }
        assert_transition(ActionStatus::WaitingApproval, ActionStatus::Approved)
            .map_err(|e| rusqlite::Error::ToSqlConversionFailure(Box::new(e)))?;
        let n2 = tx.execute(
            "UPDATE execution_records SET status = 'APPROVED', approved_by = ?1
             WHERE execution_id = ?2 AND status = 'WAITING_APPROVAL'",
            params![decided_by, execution_id],
        )?;
        if n2 != 1 {
            return Err(rusqlite::Error::ToSqlConversionFailure(Box::new(SqubeGuardError(
                "execution transition failed".into(),
            ))));
        }
        tx.commit()?;
        Ok(execution_id)
    }

    pub fn consume_approval(
        &self,
        approval_id: &str,
        execution_id: &str,
        now_iso: &str,
    ) -> rusqlite::Result<()> {
        let tx = self.conn.unchecked_transaction()?;
        let status: String = tx.query_row(
            "SELECT status FROM approval_requests WHERE approval_id = ?1 AND execution_id = ?2",
            params![approval_id, execution_id],
            |row| row.get(0),
        )?;
        if status == "CONSUMED" {
            return Err(rusqlite::Error::ToSqlConversionFailure(Box::new(SqubeApprovalError {
                approval_id: approval_id.to_string(),
                reason: "already consumed".into(),
            })));
        }
        if status != "GRANTED" {
            return Err(rusqlite::Error::ToSqlConversionFailure(Box::new(SqubeApprovalError {
                approval_id: approval_id.to_string(),
                reason: format!("status is {}", status),
            })));
        }
        let n = tx.execute(
            "UPDATE approval_requests SET status = 'CONSUMED', decided_at = ?1
             WHERE approval_id = ?2 AND status = 'GRANTED'",
            params![now_iso, approval_id],
        )?;
        if n != 1 {
            return Err(rusqlite::Error::ToSqlConversionFailure(Box::new(SqubeApprovalError {
                approval_id: approval_id.to_string(),
                reason: "concurrent consume lost race".into(),
            })));
        }
        assert_transition(ActionStatus::Approved, ActionStatus::Executing)
            .map_err(|e| rusqlite::Error::ToSqlConversionFailure(Box::new(e)))?;
        let n2 = tx.execute(
            "UPDATE execution_records SET status = 'EXECUTING' WHERE execution_id = ?1 AND status = 'APPROVED'",
            params![execution_id],
        )?;
        if n2 != 1 {
            return Err(rusqlite::Error::ToSqlConversionFailure(Box::new(SqubeGuardError(
                "execution transition failed".into(),
            ))));
        }
        tx.commit()?;
        Ok(())
    }

    pub fn latest_status(&self) -> rusqlite::Result<Option<String>> {
        let mut stmt = self
            .conn
            .prepare("SELECT status FROM execution_records ORDER BY created_at DESC LIMIT 1")?;
        let mut rows = stmt.query([])?;
        if let Some(row) = rows.next()? {
            Ok(Some(row.get(0)?))
        } else {
            Ok(None)
        }
    }
}
